metadata-ingestion/src/datahub/ingestion/source/unity/ge_profiler.py

```diff
@@ -59,51 +60,100 @@ def __init__(
+        else:
+            url = self.config.get_sql_alchemy_url()
+
+        logger.debug(f"sql_alchemy_url={url}")
+
+        engine = create_engine(url, **self.config.options)
+        with engine.connect() as conn:
+            inspector = inspect(conn)
+
+        return DatahubGEProfiler(
+            conn=inspector.bind,
+            report=self.report,
+            config=self.config.profiling,
+            platform=self.platform,
+            env=self.config.env,
+        )
+
     def get_workunits(self, tables: List[Table]) -> Iterable[MetadataWorkUnit]:
         # Extra default SQLAlchemy option for better connection pooling and threading.
         # https://docs.sqlalchemy.org/en/14/core/pooling.html#sqlalchemy.pool.QueuePool.params.max_overflow
         self.config.options.setdefault(
             "max_overflow", self.profiling_config.max_workers
         )
 
-        url = self.config.get_sql_alchemy_url()
-        engine = create_engine(url, **self.config.options)
-        conn = engine.connect()
-
-        profile_requests = []
-        with ThreadPoolExecutor(
-            max_workers=self.profiling_config.max_workers
-        ) as executor:
-            futures = [
-                executor.submit(
-                    self.get_unity_profile_request,
-                    UnityCatalogSQLGenericTable(table),
-                    conn,
-                )
-                for table in tables
-            ]
+        # Group tables by catalog to create connections for each catalog
+        tables_by_catalog = {}
+        for table in tables:
+            catalog = table.ref.catalog
+            if catalog not in tables_by_catalog:
+                tables_by_catalog[catalog] = []
+            tables_by_catalog[catalog].append(table)
 
-            try:
-                for i, completed in enumerate(
-                    as_completed(futures, timeout=self.profiling_config.max_wait_secs)
-                ):
-                    profile_request = completed.result()
-                    if profile_request is not None:
-                        profile_requests.append(profile_request)
-                    if i > 0 and i % 100 == 0:
-                        logger.info(f"Finished table-level profiling for {i} tables")
-            except (TimeoutError, concurrent.futures.TimeoutError):
-                logger.warning("Timed out waiting to complete table-level profiling.")
-
-        if len(profile_requests) == 0:
-            return
-
-        yield from self.generate_profile_workunits(
-            profile_requests,
-            max_workers=self.config.profiling.max_workers,
-            platform=self.platform,
-            profiler_args=self.get_profile_args(),
-        )
+        # Process each catalog separately to ensure proper database context
+        for catalog, catalog_tables in tables_by_catalog.items():
+            # Create connection for this specific catalog
+            url = self.config.get_sql_alchemy_url(database=catalog)
```

Comment:
I think it might make sense to create a `_get_workunits_for_catalog(catalog_name, catalog_tables)` helper method